source-mongodb/pull.go

```diff
@@ -0,0 +1,267 @@
+		// connector will run a backfill.
+		if e, ok := err.(mongo.ServerError); ok {
+			if e.HasErrorCode(changeStreamFatalErrorCode) && e.HasErrorMessage(resumePointGoneErrorMessage) {
+				if err = c.Output.Checkpoint([]byte("{}"), false); err != nil {
+					return fmt.Errorf("output checkpoint failed: %w", err)
+				}
+
+				return fmt.Errorf("change stream on collection %s cannot resume capture, the connector will restart and run a backfill: %w", res.Collection, err)
+			}
+		}
+
+		return fmt.Errorf("change stream on collection %s: %w", res.Collection, err)
+	}
+
+	for cursor.Next(ctx) {
+		var ev changeEvent
+		if err = cursor.Decode(&ev); err != nil {
+			return fmt.Errorf("decoding document in collection %s: %w", res.Collection, err)
+		}
+
+		var doc = sanitizeDocument(ev.FullDocument)
+
+		js, err := json.Marshal(doc)
+		if err != nil {
+			return fmt.Errorf("encoding document in collection %s as json: %w", res.Collection, err)
+		}
+
+		if err = c.Output.Documents(binding, js); err != nil {
+			return fmt.Errorf("output documents failed: %w", err)
+		}
+
+		var checkpoint = captureState{
+			Resources: map[string]bson.Raw{
+				resourceId(res): cursor.ResumeToken(),
+			},
+		}
+
+		checkpointJson, err := json.Marshal(checkpoint)
+		if err != nil {
+			return fmt.Errorf("encoding checkpoint to json failed: %w", err)
+		}
+
+		if err = c.Output.Checkpoint(checkpointJson, true); err != nil {
+			return fmt.Errorf("output checkpoint failed: %w", err)
+		}
+	}
+
+	if err := cursor.Err(); err != nil {
+		return fmt.Errorf("cursor error for change stream %s: %w", res.Collection, err)
+	}
+
+	defer cursor.Close(ctx)
+
+	return nil
+}
+
+const CHECKPOINT_EVERY = 100
+
+func (c *capture) BackfillCollection(ctx context.Context, client *mongo.Client, binding uint32, res resource) error {
+	var db = client.Database(res.Database)
+
+	var collection = db.Collection(res.Collection)
+
+	log.Debug("finding documents in collection ", res.Collection)
+	cursor, err := collection.Find(ctx, bson.D{})
+	if err != nil {
+		return fmt.Errorf("could not run find query on collection %s: %w", res.Collection, err)
+	}
+
+	defer cursor.Close(ctx)
+
+	var i = 0
+	for cursor.Next(ctx) {
+		i += 1
+		var doc bson.M
+		if err = cursor.Decode(&doc); err != nil {
+			return fmt.Errorf("decoding document in collection %s: %w", res.Collection, err)
+		}
+		doc = sanitizeDocument(doc)
+
+		js, err := json.Marshal(doc)
+		if err != nil {
+			return fmt.Errorf("encoding document in collection %s as json: %w", res.Collection, err)
+		}
+
+		if err = c.Output.Documents(binding, js); err != nil {
+			return fmt.Errorf("output documents failed: %w", err)
+		}
+
+		if i % CHECKPOINT_EVERY == 0 {
+			if err = c.Output.Checkpoint([]byte("{}"), true); err != nil {
+				return fmt.Errorf("output checkpoint failed: %w", err)
+			}
+		}
+	}
+
+	// minus five seconds for potential error
+	c.BackfillFinishedAt = &primitive.Timestamp{ T: uint32(time.Now().Unix()) - 5, I: 0 }
```

Comment:
It feels like this timestamp value ought to be a Go `time.Time` here (something like `time.Now().Sub(backfillOverlapDuration)`) which gets converted to a BSON `primitive.Timestamp` in the `SetStartAtOperationTime` part of the logic.